aten/src/THC/THCTensorMathPointwise.cuh

```diff
@@ -779,5 +779,49 @@ struct TensorBitXorOp {
   }
 };
 
+template <typename real, typename accreal>
+struct TensorDigammaOp {
```

Comment:
@zou3519 could you help me get this CUDA version working? I tried to translate the working C version, but `test/test_cuda.py::TestCuda::test_FloatTensor_digamma` is failing and I think I may have missed some subtleties of CUDA.
<details>

```
_________________________________________ TestCuda.test_FloatTensor_digamma __________________________________________

self = <test_cuda.TestCuda testMethod=test_FloatTensor_digamma>

    def tmp(self):
        cpu_tensor = tensor_constructor(t)
        type_map = {}
        if force_gpu_half:
            type_map = {
                'torch.FloatTensor': 'torch.cuda.HalfTensor',
                'torch.DoubleTensor': 'torch.cuda.HalfTensor',
            }
        gpu_tensor = to_gpu(cpu_tensor, type_map)
        cpu_args = arg_constructor(t)
        gpu_args = [to_gpu(arg, type_map) for arg in cpu_args]
        cpu_result = getattr(cpu_tensor, fn)(*cpu_args)
        try:
            gpu_result = getattr(gpu_tensor, fn)(*gpu_args)
        except RuntimeError as e:
            reason = e.args[0]
            if 'only supports floating-point types' in reason or 'unimplemented data type' in reason:
                raise unittest.SkipTest('unimplemented data type')
            raise
        except AttributeError as e:
            reason = e.args[0]
            if 'object has no attribute' in reason:
                raise unittest.SkipTest('unimplemented data type')
            raise
        # If one changes, another should change as well
        self.assertEqual(cpu_tensor, gpu_tensor, precision)
        self.assertEqual(cpu_args, gpu_args, precision)
        # Compare results
>       self.assertEqual(cpu_result, gpu_result, precision)

test/test_cuda.py:421:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test/common.py:228: in assertEqual
    assertTensorsEqual(x, y)
test/common.py:220: in assertTensorsEqual
    self.assertLessEqual(max_err, prec, message)
E   AssertionError: 0.185546875 not less than or equal to 1e-05
```
</details>

Reply:
Never mind, this was just the same issue you had with large values causing large absolute error but low relative error.